### PR TITLE
Feature proposal: check MX on email address domains

### PR DESCRIPTION
**Pull request by rfk**
From @ryanfeeley in https://github.com/mozilla/fxa-content-server/issues/3500.  We'll need to figure out the details of where and how to do the proposed MX record check.

### CODE REVIEW

**Comment by jrgm** [comment]
There is a configuration option in `postfix`, `reject_unknown_recipient_domain` to error if there is no MX or A record on the recipient DNS domain. However, there is some inconsistency in handling of errors from the `postfix` local handoff that we should address. Aside from addressing those, we may want finer control of the error handling by implementing the DNS lookup in nodejs (and, of course, consider the production implications of these lookups. For one, caching commonly valid DNS hits (perhaps `nscd(8)` can help, but I haven't checked if that caches MX lookups)).

My caveats aside, I do like the idea of failing faster on unknown smtp recipient domains. We would pick that up from the bounce queue, but I think the user experience would be better by failing faster.

**Comment by rfk** [comment]
I also like the idea of doing this synchronously somehow, rather than via the bounce queue.  Two possible options:

1. Have /account/create check for an MX record before it tries to send the email, and return an error if one is not found.
2. Add a new endpoint specifically for checking the status of an email address.

Doing (2) would let us check it as soon as the user completes the email field, potentially saving them the trouble of submitting the form with invalid data.

**Comment by jrgm** [comment]
> Have /account/create check for an MX record before it tries to send the email, and return an error if one is not found.

One small protocol nit: For completeness, the check should be for an MX on the domain, and failing that, then for an A record on the domain. But we could just check for MX and not worry about the latter case.

**Comment by rfk** [comment]
As part of this, I'm going to look at making our handling of postfix errors more consistent, which might actually make this cleaner overall.

**Comment by rfk** [comment]
> However, there is some inconsistency in handling of errors from the postfix local
> handoff that we should address.

Concretely, I tested this on latest by signing up for an account "testme@localhost".  The initial account creation is allowed to proceed to the "confirm your account" page.  But on that page, if I click "resend email", it gives me a 500 status response and an "Unexpected Error" message.  So it looks like the resend-the-email endpoint handles synchronous delivery errors by erroring out, while the /account/create endpoint silently ignores them.

**Comment by rfk** [comment]
> There is a configuration option in postfix, reject_unknown_recipient_domain to error if there
> is no MX or A record on the recipient DNS domain.

I wasn't able to make this work on latest.dev, tried setting a few variants of:

```
smtpd_recipient_restrictions = permit_mynetworks, reject_unknown_recipient_domain, reject_unauth_destination
```

But emails to things like "example@mytld" were still accepted and sent for delivery.  @jrgm can you offer any advice on enabling this in dev to test it out?

**Comment by jrgm** [comment]
Hmm.

I have
```
smtpd_recipient_restrictions =
  reject_unknown_recipient_domain
  permit_mynetworks
  reject_unauth_destination
```
and I get `NOQUEUE: reject: RCPT from localhost[127.0.0.1]: 450 4.1.2 <abcabdab@ghghhghghghgh>: Recipient address rejected: Domain not found; from=<verification@jrgm.dev.lcip.org> to=<abcabdab@ghghhghghghgh> proto=ESMTP helo=<[127.0.0.1]>`

**Comment by jrgm** [comment]
However, the postfix error is not handed by to the frontend as an error.

**Comment by rfk** [comment]
Looks like the ordering is significant in that directive, I copied what @jrgm has above and nonsense bouncey emails were synchronously rejected by postfix.  So, I think I have enough to move forward with the "TBD" aspects of defining this.

**Comment by rfk** [comment]
I think this is shaping up to be a pretty nice feature, straightforward to implement with clear user benefit.  @ryanfeeley could you please have a read-through to see if it covers everything from your point of view?  @shane-tomlinson r? on the implementation strategy?

**Comment by ryanfeeley** [comment]
@rfk Looks good to me, however if there are false positives are we locking out potential users? Should we add a way to let users report false positives? Modify the error to be something like:

* Mistyped email? It does not look like domain.com offers email. **Are we wrong about this?**

**Comment by rfk** [comment]
One of the nice things about using postfix to make the decision here, is I don't think there's a chance of any false positives - if postfix errors out, we are definitely not going to be able to send email to that address.

@ryanfeeley maybe we can re-word the error in the other direction, to say there was *definitely* a problem with sending email to that address?

File: features/proposed/check-domain-mx/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+<img src="no-mx-error.png">
+
+## Success Criteria
+
+We will measure success
+by measuring the rate of email bounces
+(`fxa.auth.account.email_bounced`)
+as a percentage of email verifications
+(`fxa.auth.account.verified`).
```
Comment (shane-tomlinson) [inline features/proposed/check-domain-mx/README.md]: What about just as a % of emails that are sent (fxa.auth.account.created)? Seems much closer to the source.

File: features/proposed/check-domain-mx/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+is around 12%:
+
+<img src="initial-bounce-rate.png">
+
+A successull outcome for this feature
+will be to reduce the bounce rate
+by 2% or more.
+
+For an additional measure of usefulness
+we will measure the number of times
```
Comment (shane-tomlinson) [inline features/proposed/check-domain-mx/README.md]: I think more interesting is the % of users that see the message. It's a proxy measure for users that either mistype an email, are being lazy and not adding the TLD, or are just trying to pull one over on us.

**Comment by shane-tomlinson** [comment]
> Mistyped email? It does not look like domain.com offers email. Are we wrong about this?

@ryanfeeley - Adding the "are we wrong about this" link implies the user can take some sort of meaningful action.

> @ryanfeeley maybe we can re-word the error in the other direction, to say there was definitely a problem with sending email to that address?

:+1: Since the domain has no A or MX record, this seems like the right path to me. If there is no A record, the domain cannot be reached at all. If there is no MX record, the domain administrator has not registered it's email provider with DNS. I feel like that is not a problem we should work around, rather that's a problem the domain administrator needs to fix.

**Comment by ryanfeeley** [comment]
@rfk @shane-tomlinson I think the original copy is still applicable, no? no-mx

**Comment by rfk** [comment]
> I think the original copy is still applicable

SGTM

File: features/proposed/check-domain-mx/README.md (hunk 1 above)
Comment (rfk) [inline features/proposed/check-domain-mx/README.md]: You caught me; there was some weirdness with the `fxa.auth.account.created` events in datadog that I couldn't figure out, so I punted.  It is indeed a better denominator so we should use it.

File: features/proposed/check-domain-mx/README.md (hunk 2 above)
Comment (rfk) [inline features/proposed/check-domain-mx/README.md]: denominator = number of successful submissions of that form?

**Comment by vladikoff** [comment]
ping @rfk

**Comment by rfk** [comment]
Merging with @shane-tomlinson's comments addressed.

**Pull request merged**